core/src/main/java/org/geysermc/geyser/extension/GeyserExtensionLoader.java

```diff
@@ -163,53 +166,62 @@ protected void loadAllExtensions(@NonNull ExtensionManager extensionManager) {
             Map<String, Path> extensions = new LinkedHashMap<>();
             Map<String, GeyserExtensionContainer> loadedExtensions = new LinkedHashMap<>();
 
-            Pattern[] extensionFilters = this.extensionFilters();
-            List<Path> extensionPaths = Files.walk(extensionsDirectory).toList();
-            extensionPaths.forEach(path -> {
-                if (Files.isDirectory(path)) {
+            Path updateDirectory = extensionsDirectory.resolve("update");
+            if (Files.isDirectory(updateDirectory)) {
+                // Get the current extensions and store them in a map
+                Map<String, List<Path>> extensionFiles = new HashMap<>();
+                this.processExtensionsFolder(extensionsDirectory, (path, description) -> extensionFiles.computeIfAbsent(description.id(), k -> new ArrayList<>()).add(path), (path, e) -> {
+                    // this file will throw again when we actually try to load extensions, and it will be handled there
+                });
+
+                // Perform the updates
+                this.processExtensionsFolder(updateDirectory, (path, description) -> {
+                    // Remove the old extension files with the same ID if it exists
+                    List<Path> oldExtensionFiles = extensionFiles.get(description.id());
+                    if (oldExtensionFiles != null) {
+                        for (Path oldExtensionFile : oldExtensionFiles) {
+                            Files.delete(oldExtensionFile);
+                        }
+                    }
+
+                    // Overwrite the extension with the new jar
+                    Files.move(path, extensionsDirectory.resolve(path.getFileName()), StandardCopyOption.REPLACE_EXISTING);
+                }, (path, e) -> {
+                    GeyserImpl.getInstance().getLogger().error(GeyserLocale.getLocaleStringLog("geyser.extensions.update.failed", path.getFileName()), e);
+                });
+            }
+
+            this.processExtensionsFolder(extensionsDirectory, (path, description) -> {
```

Comment:
let's add a comment here to explain the step; something like `3. Load the extensions`